### PR TITLE
[GDN] Amortize ReplaySSM checkpoint materialization

### PR DESCRIPTION
**Pull request by YAMY1234**
## Motivation

ReplaySSM speculative decoding currently materializes the full GDN checkpoint state on every target step, even though most speculative iterations only append a small amount of history. Deferring that work removes repeated state materialization from the decode critical path.

## Modifications

- Store speculative GDN updates in a bounded circular history and materialize the checkpoint once per accepted verify window.
- Preserve recurrent BF16 behavior with compensated high/low representations for the compact update vector and normalized key.
- Keep replay history and cursors request-scoped while keeping the physical checkpoint Mamba-slot-scoped.
- Size ReplaySSM scratch by active request capacity, and preserve reset, donation, retraction, and recovery lifetimes.

The review scope is limited to the ReplaySSM kernel, the request/state ownership and accounting needed by that kernel, and one existing accounting test. The larger experimental compatibility framework and additional defensive abstractions are not part of this PR.

## Accuracy Tests

- Nine targeted GPU tests passed, covering existing accounting/support checks plus BF16 and FP32 64-window recurrent comparisons and request-capacity allocation.
- A monolithic aggregate TP4 serving smoke test completed successfully and returned exactly 512 output tokens.
- ReplaySSM scratch is `0.857 GB/GPU` versus `0.849 GB/GPU` before the PR. CUDA-graph startup available memory is `36.129-36.131 GB` versus `36.143-36.145 GB`; the earlier approximately `3.6 GB/GPU` headroom regression is removed.

### AIME25 128K x4

The comparison uses Qwen3.5 on one 4-GPU node, aggregate TP4/DP1/PP1/EP1 serving, MTP, ReplaySSM, BF16 SSM state, 128K context, four repeats, and the same 30 problems.

| Configuration | ReplaySSM | pass@1 | majority@4 | pass@4 |
| --- | --- | ---: | ---: | ---: |
| Before PR: old ReplaySSM implementation | Enabled | 86.67% | 88.33% | **96.67% (29/30)** |
| Current ReplaySSM implementation | Enabled | 87.50% | 96.11% | **100.00% (30/30)** |

| Configuration | Repeat 0 | Repeat 1 | Repeat 2 | Repeat 3 | Total |
| --- | ---: | ---: | ---: | ---: | ---: |
| Before PR: old ReplaySSM implementation | 28/30 | 26/30 | 25/30 | 25/30 | 104/120 |
| Current ReplaySSM implementation | 26/30 | 27/30 | 26/30 | 26/30 | 105/120 |

The current run produced all `120/120` predictions with zero harness errors and `3,019,855` completion tokens.

### AIME25 128K x8

For additional confidence, the old and current ReplaySSM implementations were rerun with the same serving/evaluation contract and eight repeats. `pass@N` is the fraction of the 30 problems with at least one correct answer among `N` attempts; it is not the average accuracy across repeats.

| Configuration | ReplaySSM | pass@1 | majority@8 | pass@8 |
| --- | --- | ---: | ---: | ---: |
| Before PR: old ReplaySSM implementation | Enabled | 91.67% | 96.67% | **100.00% (30/30)** |
| Current ReplaySSM implementation | Enabled | 92.08% | 97.78% | **100.00% (30/30)** |

| Configuration | R0 | R1 | R2 | R3 | R4 | R5 | R6 | R7 | Total |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| Before PR: old ReplaySSM implementation | 28/30 | 26/30 | 28/30 | 29/30 | 28/30 | 27/30 | 27/30 | 27/30 | 220/240 |
| Current ReplaySSM implementation | 26/30 | 26/30 | 28/30 | 28/30 | 27/30 | 29/30 | 28/30 | 29/30 | 221/240 |

Both arms produced all `240/240` predictions with zero harness errors and passed all 30 problems at pass@8. The current implementation is `+1/240` in total correct attempts (`92.08%` versus `91.67%`).

### Logits and long-sequence stability

- Four independent AIME27 traces completed successfully; all four produced the correct answer (`248`).
- Over the token-aligned common prefix with the old implementation (406 tokens on average), top-1 agreement is `99.92%`, mean top-5 Jensen-Shannon divergence is `3.82e-4`, and mean absolute chosen-logprob delta is `7.42e-3`.
- An independent rerun of the compact implementation gives `100%` top-1 agreement and a `7.61e-3` mean absolute chosen-logprob delta over its common prefix, placing the old-versus-current delta at the same scale as run-to-run variation.
- After token paths diverge, the values are confidence diagnostics rather than token-aligned numerical error. Across a 34,816-token horizon, chosen-logprob variance decreases from `0.0194` to `0.0105` and entropy decreases from `0.0706` to `0.0464`; there is no evidence of error or variance accumulating with generation length.

## Speed Tests and Profiling

- Setup: Qwen3.5-397B-A17B-NVFP4 on one node with 4 GB300 GPUs in DP4, MTP with 6 draft tokens, TRTLLM-MHA split 1, ReplaySSM `extra_buffer`, radix-cache page size 64, and the same frozen 112-request cohort in both arms.
- Workload: every request uses exactly 88,737 input tokens and generates exactly 8,192 output tokens. Requests are prefetched before measurement; no new prefill is admitted during the pure-generation interval.
- Acceptance criteria: 112/112 HTTP 200 responses, routed distribution `[28, 28, 28, 28]`, achieved batch 28 on every rank, CUDA graph enabled, queue depth 0, no replenishment, and measurement ending before the first completion.

| Variant | Natural mean acceptance length | Mean context/request | Target-step time (ms) | Fixed-AL 4.8 output tokens/s/GPU | Change vs. baseline | Observed output tokens/s/GPU | Change vs. baseline |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| Before PR: old ReplaySSM implementation | 4.32726 | 6,311.26 | 39.74513 | 3,381.622 | - | 3,048.695 | - |
| Current ReplaySSM implementation | 4.27298 | 6,201.45 | 37.94529 | 3,541.988 | **+4.742%** | 3,153.163 | **+3.427%** |

- The current implementation reduces target-step time by `4.528%` versus the baseline.
- Natural acceptance length changes by `-1.254%` and mean context by `-1.740%`; the fixed-AL metric normalizes acceptance to 4.8, while observed throughput remains `+3.427%`.
- Rank CV for fixed-AL throughput is below 0.6% in both arms.
- `Output tokens/s/GPU` is an engine-side pure-generation metric from the fixed-B28 target-step window, not a service-level AIPerf total-throughput metric.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #33840060530](https://github.com/sgl-project/sglang/actions/runs/33840060530)
Latest PR Test (Extra): :x: [Run #33840060375](https://github.com/sgl-project/sglang/actions/runs/33840060375)
Latest PR Test (AMD ROCm 7.2): :hourglass_flowing_sand: [Run #33840060550](https://github.com/sgl-project/sglang/actions/runs/33840060550)

### CODE REVIEW

**Comment by YAMY1234** [comment]
/rerun-test test/registered/models_e2e/test_qwen35_fp4_mtp.py::TestQwen35FP4MTPReplaySSM.test_gsm8k test/registered/attention/unittests/gdn/test_gdn_replayssm_spec_fold.py test/registered/attention/unittests/gdn/test_gdn_cutedsl_ring_verify.py test/registered/attention/unittests/gdn/test_linear_replayssm_decode.py

**Comment by YAMY1234** [comment]
Added the Replay SSM AIME results. There is no accuracy regression.
image

File: python/sglang/kernels/ops/attention/fla/gdn_replayssm_spec_decode.py [hunk 1]
```diff
@@ -277,12 +284,11 @@ def gdn_replayssm_spec_circular_kernel(
             mask=mask_kt[None, :],
             other=0.0,
         ).to(tl.float32)
-        # Keep the raw (pre-norm) key tile for the exact-fold ring: the replay
-        # recomputes the L2 norm in fp32 exactly as the recurrent kernel does
-        # (its division form differs bitwise from the reciprocal-multiply below).
         k_raw_tile = k_tile
-        q_tile = (q_tile * (q_rnorm * scale)[:, None]).to(h0.dtype.element_ty)
-        k_tile = (k_tile * k_rnorm[:, None]).to(h0.dtype.element_ty)
+        q_tile = (q_tile * (q_rnorm * scale)[:, None]).to(q.dtype.element_ty)
```
Comment (Qiaolin-Yu) [inline python/sglang/kernels/ops/attention/fla/gdn_replayssm_spec_decode.py]: we downgrade the precision of q_tile/k_tile, b_d_scaled, and khist_tile, right? Is this expected?

**Comment by Qiaolin-Yu** [comment]
> Added the Replay SSM AIME results. There is no accuracy regression. image
> 
could we run more pass? maybe repeat 8 or repeat 16. Does the pass@4 mean average of 4 runs? It seems the pass@4 is higher but total is lower?

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 2]
```diff
@@ -717,6 +707,19 @@ def __init__(
                         dtype=torch.float32,
                         device=device,
                     )
+                elif enable_linear_replayssm_spec and ssm_dtype != torch.float32:
+                    # Low parts of compact D and normalized K. A BF16 checkpoint
+                    # materializes these once per accepted verify window.
+                    replayssm_rawv = torch.zeros(
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/mem_cache/memory_pool.py:708]: why do we hold low parts for ssm_dtype==bf16 but not for ssm_dtype==fp32?

File: python/sglang/kernels/ops/attention/fla/gdn_replayssm_spec_decode.py
Comment (Qiaolin-Yu) [inline python/sglang/kernels/ops/attention/fla/gdn_replayssm_spec_decode.py:970]: should we use the same condition as https://github.com/sgl-project/sglang/pull/35544/changes#diff-4f7380584c282184fbe2f4c7e536566bd9ac10764d0a527125070d63ffc65a59R934?

File: python/sglang/kernels/ops/attention/fla/gdn_replayssm_spec_decode.py (hunk 1 above)
Comment (YAMY1234) [reply python/sglang/kernels/ops/attention/fla/gdn_replayssm_spec_decode.py]: Good catch. The FP32-checkpoint path was unintentionally lowering these operands; `9168a3f7dd` restores its FP32/TF32 path while leaving the BF16 hot path unchanged. The BF16 and FP32 64-window GPU comparisons both pass.

File: python/sglang/srt/mem_cache/memory_pool.py (hunk 2 above)
Comment (YAMY1234) [reply python/sglang/srt/mem_cache/memory_pool.py:708]: The low parts should follow the compact ring dtype, not the checkpoint dtype, because spec-verify `replayssm_d/k` use the activation/conv dtype. `9168a3f7dd` changes allocation and accounting to `ring_dtype != torch.float32`, including the FP32-checkpoint + BF16-ring case.

File: python/sglang/kernels/ops/attention/fla/gdn_replayssm_spec_decode.py
Comment (YAMY1234) [reply python/sglang/kernels/ops/attention/fla/gdn_replayssm_spec_decode.py:970]: These two kernels already use the same condition: `row_mask & (index > NULL_BLOCK_ID)`. This path passes `NULL_BLOCK_ID=-1`, so slots `>= 0` are valid and `-1` is padding; no code change is needed.

**Comment by YAMY1234** [comment]
`pass@4`/`pass@8` mean the fraction of the 30 problems with at least one correct answer among 4/8 attempts, not the average repeat accuracy. I ran the matched 128K x8 comparison: old ReplaySSM was `220/240` (`91.67%`) and current was `221/240` (`92.08%`); both completed `240/240` predictions with zero harness errors and reached pass@8 `100% (30/30)`. The PR description now includes both x4 and x8 tables.

File: python/sglang/kernels/ops/attention/fla/gdn_replayssm_spec_decode.py
Comment (Qiaolin-Yu) [reply python/sglang/kernels/ops/attention/fla/gdn_replayssm_spec_decode.py:970]: https://github.com/sgl-project/sglang/pull/35544/changes#diff-4f7380584c282184fbe2f4c7e536566bd9ac10764d0a527125070d63ffc65a59L983. Though the fifth param of `_advance_gdn_spec_cursors_kernel` is `state_batch_indices_ptr`, it actually pass `replay_indices` in?

Comment (YAMY1234) [reply python/sglang/kernels/ops/attention/fla/gdn_replayssm_spec_decode.py:970]: Yes—the cursors are request-slot keyed, so this helper intentionally receives `replay_indices`; the kernel parameter name was stale. I renamed it and its stride in `7afa1bea2b`; `null_block_id=-1` still keeps request slots `>= 0` valid.

**Review by Qiaolin-Yu — APPROVED** [review]

**Comment by YAMY1234** [comment]
All NV CI passed https://github.com/sgl-project/sglang/actions/runs/33840060530/job/101168492512?pr=35544
